org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/XMLFormatterDocumentNew.java

```diff
@@ -368,6 +389,20 @@ boolean hasLineBreak(int startAttr, int start) {
 		return false;
 	}
 
+	public int getNormalizedLengthAtRange(int startline, int initialOffset) {
+		String text = textDocument.getText().substring(startline, initialOffset);
+		int contentOffset = 0;

```

Comment:
Please dont use substring which can be bad for performance memory.

I think you czn do the same code without substring.